### ISSUE TITLE
Error code confusion in mocked responses?

### ISSUE THREAD

**Issue by okomestudio** [spulec/moto#1004]
It appears an error response including `Code` and `Message` is rendered using `moto.core.exceptions.ERROR_RESPONSE`, which is passed to the template via `RESTError(error_type, message)` where `error_type` gets rendered into `Code` of the template.

In `exceptions.py` of service submodules, `error_type` often is a short descriptive text, instead of HTTP status code which appears to be the case in many non-mocked responses. (In fact, real responses are indeed inconsistent in `Code` being a numeric status code vs. short text...)

Unfortunate side effect of the difference is that an idiomatic code like the following does not work properly (this is from a `boto3` example page):

```
import botocore
bucket = s3.Bucket('mybucket')
exists = True
try:
    s3.meta.client.head_bucket(Bucket='mybucket')
except botocore.exceptions.ClientError as e:
    # If a client error is thrown, then check that it was a 404 error.
    # If it was a 404 error, then the bucket does not exist.
    error_code = int(e.response['Error']['Code'])
    if error_code == 404:
        exists = False
```

since `e.response['Error']['Code']` of mocked response, being a text, cannot be cast properly to `int`.

I'm wondering if this is a design decision of some sort of oversight? Is there a reason why `Code` and `Message` are not mocked closely to the actual responses?

**Comment by JackDanger** [comment]
@okomestudio can you share what the `e.response['Error']['Code']` is from that example? Is it 'NoSuchBucket'? If so, I'm confused why the boto3 docs suggest that would work, I think it wouldn't.

**Comment by okomestudio** [comment]
The specific example snippet is from this page: http://boto3.readthedocs.io/en/latest/guide/migrations3.html

I think the return value of `response['Error']['Code']` might be inconsistently implemented (numeric code vs. `NoSuchKey`) in the AWS API itself, and I initially found this out when I recently updated `moto` to the recent version and a couple existing tests (on EMR resources) started failing due to this.

This might be one of those things we have to discover by actually hitting the API and inspecting responses manually.

**Comment by JackDanger** [comment]
We recently changed this very response code in Moto. When you issue a GET
request you get NoSuchKey and when you issue a HEAD request you get a 404.
Or perhaps that's backwards. Regardless, it was a bug where a user found
the AWS API used different responses for those two HTTP verbs so we matched
the live AWS response.
Il giorno lun 17 lug 2017 alle 5:35 PM Taro Sato <notifications@github.com>
ha scritto:
